### ISSUE TITLE
GrpcClientTest.maxOutboundSize_tooBig is flaky?

### ISSUE THREAD

**Issue by imasahiro** [line/armeria#822]
Didn't dig to the detail but I saw the test failure at https://travis-ci.org/line/armeria/builds/297503513 (https://github.com/line/armeria/pull/819)

```
GrpcClientTest.maxOutboundSize_tooBig: FAILURE

----------------------------------------------

10:47:35.619 [armeria-eventloop-epoll-5-1] WARN  c.l.a.server.HttpResponseSubscriber - [id: 0x01242a07, L:/127.0.0.1:37549 - R:/127.0.0.1:58208] Unexpected exception from a service or a response publisher: FunctionalDecoratingService(EchoRequestHeadersInTrailers(com.linecorp.armeria.server.grpc.GrpcService@f661a58), com.linecorp.armeria.client.grpc.GrpcClientTest$1$$Lambda$53/2034608021@47656120)

io.grpc.StatusException: CANCELLED
    at io.grpc.Status.asException(Status.java:534)
    at com.linecorp.armeria.server.grpc.ArmeriaServerCall.closeListener(ArmeriaServerCall.java:362)
    at com.linecorp.armeria.server.grpc.ArmeriaServerCall.transportReportStatus(ArmeriaServerCall.java:328)
    at com.linecorp.armeria.internal.grpc.HttpStreamReader.onError(HttpStreamReader.java:134)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage$CloseEvent.notifySubscriber(DefaultStreamMessage.java:638)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberWithCloseEvent(DefaultStreamMessage.java:403)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:317)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$notifySubscriber$2(DefaultStreamMessage.java:298)
    at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:309)
    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
    at java.lang.Thread.run(Thread.java:748)

com.linecorp.armeria.client.grpc.GrpcClientTest > maxOutboundSize_tooBig FAILED
    java.lang.AssertionError:
    Expecting:
      <java.lang.IllegalStateException: ClientCall already closed>
    to be an instance of:
      <io.grpc.StatusRuntimeException>
    but was:
      <"java.lang.IllegalStateException: ClientCall already closed
        at com.google.common.base.Preconditions.checkState(Preconditions.java:501)
        at io.grpc.stub.ClientCalls$BlockingResponseStream$QueuingListener.onClose(ClientCalls.java:566)
        at com.linecorp.armeria.client.grpc.ArmeriaClientCall.cancel(ArmeriaClientCall.java:182)
        at com.linecorp.armeria.client.grpc.ArmeriaClientCall.sendMessage(ArmeriaClientCall.java:209)
        at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:237)
        at io.grpc.stub.ClientCalls.blockingServerStreamingCall(ClientCalls.java:156)
        at com.linecorp.armeria.grpc.testing.TestServiceGrpc$TestServiceBlockingStub.streamingOutputCall(TestServiceGrpc.java:448)
        at com.linecorp.armeria.client.grpc.GrpcClientTest.lambda$maxOutboundSize_tooBig$5(GrpcClientTest.java:776)
        at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:60)
        at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:713)
        at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1135)
        at com.linecorp.armeria.client.grpc.GrpcClientTest.maxOutboundSize_tooBig(GrpcClientTest.java:776)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:748)
    ">
```

**Comment by trustin** [comment]
Encountered this just today at master. Any clues, @anuraaga ?

**Comment by trustin** [comment]
```
GrpcClientTest.maxOutboundSize_tooBig: FAILURE
----------------------------------------------
06:12:28.218 [armeria-common-worker-nio-3-1] WARN  c.l.a.client.logging.LoggingClient - Response: {startTime=2017-11-30T06:12:28.202Z(1512022348202), length=0B, duration=800ns, cause=io.grpc.StatusException: CANCELLED, headers=[], content=DefaultRpcResponse{cause=io.grpc.StatusException: CANCELLED}}
io.grpc.StatusException: CANCELLED
	at io.grpc.Status.asException(Status.java:534)
	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.cancel(ArmeriaClientCall.java:178)
	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.sendMessage(ArmeriaClientCall.java:203)
	at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:269)
	at io.grpc.stub.ClientCalls.blockingServerStreamingCall(ClientCalls.java:164)
	at com.linecorp.armeria.grpc.testing.TestServiceGrpc$TestServiceBlockingStub.streamingOutputCall(TestServiceGrpc.java:581)
	at com.linecorp.armeria.client.grpc.GrpcClientTest.lambda$maxOutboundSize_tooBig$5(GrpcClientTest.java:779)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:60)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:713)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1135)
	at com.linecorp.armeria.client.grpc.GrpcClientTest.maxOutboundSize_tooBig(GrpcClientTest.java:779)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Failed to frame message
	at io.grpc.Status.asRuntimeException(Status.java:517)
	at com.linecorp.armeria.internal.grpc.ArmeriaMessageFramer.writePayload(ArmeriaMessageFramer.java:125)
	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.sendMessage(ArmeriaClientCall.java:194)
	... 20 common frames omitted
Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: message too large 4 > 3
	at io.grpc.Status.asRuntimeException(Status.java:517)
	at com.linecorp.armeria.internal.grpc.ArmeriaMessageFramer.writeUncompressed(ArmeriaMessageFramer.java:169)
	at com.linecorp.armeria.internal.grpc.ArmeriaMessageFramer.writePayload(ArmeriaMessageFramer.java:108)
	... 21 common frames omitted
06:12:28.233 [armeria-eventloop-nio-5-1] WARN  c.l.a.server.HttpResponseSubscriber - [id: 0xcb843fd6, L:/127.0.0.1:1564 - R:/127.0.0.1:1565] Unexpected exception from a service or a response publisher: FunctionalDecoratingService(EchoRequestHeadersInTrailers(com.linecorp.armeria.server.grpc.GrpcService@638ee2d2), com.linecorp.armeria.client.grpc.GrpcClientTest$1$$Lambda$51/586322482@f1b15cc)
io.grpc.StatusException: CANCELLED
	at io.grpc.Status.asException(Status.java:534)
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.closeListener(ArmeriaServerCall.java:362)
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.transportReportStatus(ArmeriaServerCall.java:328)
	at com.linecorp.armeria.internal.grpc.HttpStreamReader.onError(HttpStreamReader.java:134)
	at com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:287)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:191)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:363)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:306)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:250)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.addObjectOrEvent(DefaultStreamMessage.java:236)
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.close(DefaultStreamMessage.java:381)
	at com.linecorp.armeria.server.Http2RequestDecoder.onRstStreamRead(Http2RequestDecoder.java:224)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onRstStreamRead(DefaultHttp2ConnectionDecoder.java:354)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readRstStreamFrame(DefaultHttp2FrameReader.java:516)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:260)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:118)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:388)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:448)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at com.linecorp.armeria.internal.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:108)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:138)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
com.linecorp.armeria.client.grpc.GrpcClientTest > maxOutboundSize_tooBig FAILED
    java.lang.AssertionError: 
    Expecting:
      <java.lang.IllegalStateException: ClientCall already closed>
    to be an instance of:
      <io.grpc.StatusRuntimeException>
    but was:
      <"java.lang.IllegalStateException: ClientCall already closed
    	at com.google.common.base.Preconditions.checkState(Preconditions.java:501)
    	at io.grpc.stub.ClientCalls$BlockingResponseStream$QueuingListener.onClose(ClientCalls.java:599)
    	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.close(ArmeriaClientCall.java:252)
    	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.cancel(ArmeriaClientCall.java:179)
    	at com.linecorp.armeria.client.grpc.ArmeriaClientCall.sendMessage(ArmeriaClientCall.java:203)
    	at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:269)
    	at io.grpc.stub.ClientCalls.blockingServerStreamingCall(ClientCalls.java:164)
    	at com.linecorp.armeria.grpc.testing.TestServiceGrpc$TestServiceBlockingStub.streamingOutputCall(TestServiceGrpc.java:581)
    	at com.linecorp.armeria.client.grpc.GrpcClientTest.lambda$maxOutboundSize_tooBig$5(GrpcClientTest.java:779)
    	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:60)
    	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:713)
    	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1135)
    	at com.linecorp.armeria.client.grpc.GrpcClientTest.maxOutboundSize_tooBig(GrpcClientTest.java:779)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
    	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
    	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    	at java.lang.Thread.run(Thread.java:748)
    ">
        at com.linecorp.armeria.client.grpc.GrpcClientTest.maxOutboundSize_tooBig(GrpcClientTest.java:780)
```

**Comment by trustin** [comment]
May be fixed via #1281
